source/common/uma_pools/disjoint_pool.cpp

```diff
@@ -0,0 +1,948 @@
+//===---------- disjoint_pool.cpp - Allocator for USM memory --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <algorithm>
+#include <array>
+#include <bitset>
+#include <cassert>
+#include <cctype>
+#include <iomanip>
+#include <limits>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+// TODO: replace with logger?
```

Comment:
Yes, but in a separate PR. We will need to align with making the same change in the adapters so things are consistent.